kvserver/closedts: add policy refresher

wenyihu6 opened the pull request:
This commit adds policy refresher which refreshes the closed
timestamp policies of leaseholder ranges based on observed
network latency periodically at the interval of
kv.closed_timestamp.policy_refresh_interval.

Part of: #59680
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142664)
<!-- Reviewable:end -->

wenyihu6 committed:
kvserver/closedts: add policy refresher

This commit adds policy refresher which refreshes the closed
timestamp policies of leaseholder ranges based on observed
network latency periodically at the interval of
kv.closed_timestamp.policy_refresh_interval.

Part of: #59680
Release note: none

arulajmani reviewed (commented):
I looked through your PR series. Overall, the structure looks great. 

The only high level thought I have is that it's unfortunate that we're joining the bucket against the span config to return the actual closed timestamp policy in `closedTimestampPolicyRLocked`. The unfortunate bit is that for ranges that have `LAG_BY_CLUSTER_SETTING` closed timestamp policy, we're also keeping track of a global reads bucket, which we're maintaining and refreshing, yet discarding.

How do you feel about checking the config and then storing the policy on the `Replica` struct. So it would replace `cachedLocality` in your subsequent PRs. If we go this route, we'll also want to refresh the closed timestamp policy upon a span config update, and not just from the `PolicyRefresher`. 

Happy to chat more about this idea over a call instead of iterating over this PR, if you want.

> Reviewed 2 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142664#-:-OMHt3F31ENh3kudJL6i:blhwl9h)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/kv/kvserver/closedts/multiregion/policy_refresher.go` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142664#-OMHu5UI5OliTbhremAR:-OMHu5UI5OliTbhremAS:bscpncr) ([raw file](https://github.com/cockroachdb/cockroach/blob/28a4dee03cee01fa09a300d2eab6c9641d59324e/pkg/kv/kvserver/closedts/multiregion/policy_refresher.go#L6)):*
> ```Go
> // included in the /LICENSE file.
> 
> package multiregion
> ```

nit: the multiregion package name seems misnamed here. Do we need a new package, or can we keep this in closedts? If we do need a new package, let's come up with a different name, as these policies apply to single range clusters too.
___
*[`pkg/kv/kvserver/closedts/multiregion/policy_refresher.go` line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142664#-OMEf16j3X15KcaczTSo:-OMEf16j3X15KcaczTSp:b-mcgke7) ([raw file](https://github.com/cockroachdb/cockroach/blob/28a4dee03cee01fa09a300d2eab6c9641d59324e/pkg/kv/kvserver/closedts/multiregion/policy_refresher.go#L22)):*
> ```Go
> )
> 
> // PolicyRefresher manages closed timestamp policies for multi-region ranges. It
> ```

nit: not just multi-region ranges, all ranges, right?
___
*[`pkg/kv/kvserver/closedts/multiregion/policy_refresher.go` line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142664#-OMEfEAQAuP5Ns9wwjkA:-OMEfEARAoKGymLZXHgR:bqy3dlm) ([raw file](https://github.com/cockroachdb/cockroach/blob/28a4dee03cee01fa09a300d2eab6c9641d59324e/pkg/kv/kvserver/closedts/multiregion/policy_refresher.go#L23)):*
> ```Go
> 
> // PolicyRefresher manages closed timestamp policies for multi-region ranges. It
> // periodically refreshes latency based range closed timestamp policies which
> ```

Should this say refresh policies based on latencies between the leaseholder and its farthest follower?
___
*[`pkg/kv/kvserver/closedts/multiregion/policy_refresher.go` line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142664#-OMEfPz6E28StpIH7NRC:-OMEfPz6E28StpIH7NRD:bqis6fj) ([raw file](https://github.com/cockroachdb/cockroach/blob/28a4dee03cee01fa09a300d2eab6c9641d59324e/pkg/kv/kvserver/closedts/multiregion/policy_refresher.go#L25)):*
> ```Go
> // periodically refreshes latency based range closed timestamp policies which
> // will used to determine how far into the future timestamps should be closed to
> // serve present follower reads.
> ```

Should we prefix this with "for ranges that want to serve global reads", as this PolicyRefresher should be responsible for setting the `LAG_BY_CLUSTER_SETTING` closed timestamp policy as well? 
___
*[`pkg/kv/kvserver/closedts/multiregion/policy_refresher.go` line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142664#-OMHsvJZ5gAwvrjMNglQ:-OMHsvJZ5gAwvrjMNglR:bvo2zo0) ([raw file](https://github.com/cockroachdb/cockroach/blob/28a4dee03cee01fa09a300d2eab6c9641d59324e/pkg/kv/kvserver/closedts/multiregion/policy_refresher.go#L65)):*
> ```Go
> // closed timestamp policy on a replica. Implemented by kvserver.Replica.
> type Replica interface {
> 	RefreshLatency(map[roachpb.NodeID]time.Duration)
> ```

Given we're refreshing the closed timestamp policy here, I wonder if this is better renamed `RefreshPolicy` instead.


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
Chatting this over slack. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142664#-:-OMIJO9U44syc1yvJq-6:bf6bp37)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/kv/kvserver/closedts/multiregion/policy_refresher.go` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142664#-OMHu5UI5OliTbhremAR:-OMIIslw0Ta3_s2EvWoy:b-hoyy1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/28a4dee03cee01fa09a300d2eab6c9641d59324e/pkg/kv/kvserver/closedts/multiregion/policy_refresher.go#L6)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: the multiregion package name seems misnamed here. Do we need a new package, or can we keep this in closedts? If we do need a new package, let's come up with a different name, as these policies apply to single range clusters too.
</blockquote></details>

Renamed it to `policyrefresher`. No strong opinions here, but having a separate package feels a bit nicer to me since it defines a different struct specific to performing the policy refresh, and it keeps the top-level `closedts` directory out.
___
*[`pkg/kv/kvserver/closedts/multiregion/policy_refresher.go` line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142664#-OMEf16j3X15KcaczTSo:-OMIItkJ6k_-duWMebTI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/28a4dee03cee01fa09a300d2eab6c9641d59324e/pkg/kv/kvserver/closedts/multiregion/policy_refresher.go#L22)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: not just multi-region ranges, all ranges, right?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/closedts/multiregion/policy_refresher.go` line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142664#-OMEfEAQAuP5Ns9wwjkA:-OMIIuzs8cOE_1H7_4ox:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/28a4dee03cee01fa09a300d2eab6c9641d59324e/pkg/kv/kvserver/closedts/multiregion/policy_refresher.go#L23)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should this say refresh policies based on latencies between the leaseholder and its farthest follower?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/closedts/multiregion/policy_refresher.go` line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142664#-OMEfPz6E28StpIH7NRC:-OMIIvbCDdGnJXsmvqhn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/28a4dee03cee01fa09a300d2eab6c9641d59324e/pkg/kv/kvserver/closedts/multiregion/policy_refresher.go#L25)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we prefix this with "for ranges that want to serve global reads", as this PolicyRefresher should be responsible for setting the `LAG_BY_CLUSTER_SETTING` closed timestamp policy as well?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/closedts/multiregion/policy_refresher.go` line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142664#-OMHsvJZ5gAwvrjMNglQ:-OMIJ0ho9Dx8hRna1EjH:b1iebdp) ([raw file](https://github.com/cockroachdb/cockroach/blob/28a4dee03cee01fa09a300d2eab6c9641d59324e/pkg/kv/kvserver/closedts/multiregion/policy_refresher.go#L65)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Given we're refreshing the closed timestamp policy here, I wonder if this is better renamed `RefreshPolicy` instead.
</blockquote></details>

Good idea, done. 


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
Closing in favor of https://github.com/cockroachdb/cockroach/pull/143625, doing two commits in a single PR.
